utils/fork-tree/src/lib.rs

```diff
@@ -190,9 +191,52 @@ impl<H, N, V> ForkTree<H, N, V> where
 	/// for at any point will likely be in one of the deepest chains (i.e. the
 	/// longest ones).
 	pub fn rebalance(&mut self) {
-		self.roots.sort_by_key(|n| Reverse(n.max_depth()));
-		for root in &mut self.roots {
-			root.rebalance();
+		use std::collections::HashMap;
+		use std::iter::FromIterator;
+
+		let mut current_max_depth = 1;
+		let mut max_depths = HashMap::new();
+		let mut stack = Vec::from_iter(self.roots.iter().map(|n| (n, 1, false)));
+
+		while let Some((node, depth, expanded)) = stack.pop() {
+			// we've reached a leaf, record the max depth for this branch
+			if node.children.is_empty() {
+				current_max_depth = depth;
+				max_depths.insert(node.hash.clone(), depth);
+				continue;
+			}
+
+			// this is not a leaf node and it has not been expanded yet
+			// we should reset the current max depth to the current depth
+			// since we will begin exploring a new branch
+			if !expanded {
+				for child in &node.children {
+					stack.push((node, depth, true));
```

Comment:
```suggestion
					// we always backtrack to the parent after exploring one branch
					// so that we can update its max depth before exploring the rest of
					// the branches
					stack.push((node, depth, true));
```